zebra-chain/src/keys/sapling.rs

```diff
@@ -3,25 +3,766 @@
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("NullifierDerivingKey")
+            .field("u", &hex::encode(self.get_u().to_bytes()))
+            .field("v", &hex::encode(self.get_v().to_bytes()))
+            .finish()
+    }
+}
+
+impl From<ProofAuthorizingKey> for NullifierDerivingKey {
+    /// Requires JubJub's _FindGroupHash^J("Zcash_H_", "")_, then uses
+    /// the resulting generator point to scalar multiply the
+    /// ProofAuthorizingKey into the new NullifierDerivingKey
+    ///
+    /// https://github.com/zcash/librustzcash/blob/master/zcash_primitives/src/group_hash.rs
+    /// https://zips.z.cash/protocol/protocol.pdf#saplingkeycomponents
+    /// https://zips.z.cash/protocol/protocol.pdf#concretegrouphashjubjub
+    fn from(nsk: ProofAuthorizingKey) -> Self {
+        // Should this point, when generated, be fixed for the rest of
+        // the protocol instance? Since this is kind of hash-and-pray, it
+        // seems it might not always return the same result?
+        let generator_point = zcash_h();
+
+        // TODO: impl Mul<ExtendedPoint> for Fr, so we can reverse
+        // this to match the math in the spec / general scalar mult
+        // notation convention.
+        Self(jubjub::AffinePoint::from(generator_point * nsk.0))
+    }
+}
+
+/// Magic human-readable strings used to identify what networks
+/// Sapling IncomingViewingKeys are associated with when
+/// encoded/decoded with bech32.
+mod ivk_hrp {
+    pub const MAINNET: &str = "zivks";
+    pub const TESTNET: &str = "zivktestsapling";
+}
+
+/// An _Incoming Viewing Key_, as described in [protocol specification
+/// §4.2.2][ps].
+///
+/// Used to decrypt incoming notes without spending them.
+///
+/// [ps]: https://zips.z.cash/protocol/protocol.pdf#saplingkeycomponents
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct IncomingViewingKey {
+    network: Network,
+    scalar: Scalar,
+}
+
+// TODO: impl a top-level to_bytes or PartialEq between this and [u8; 32]
+
+// TODO: impl a From that accepts a Network?
+
+impl Deref for IncomingViewingKey {
+    type Target = Scalar;
+
+    fn deref(&self) -> &Scalar {
+        &self.scalar
+    }
+}
+
+impl From<[u8; 32]> for IncomingViewingKey {
+    /// Generate an _IncomingViewingKey_ from existing bytes.
+    fn from(mut bytes: [u8; 32]) -> Self {
+        // Drop the most significant five bits, so it can be interpreted
+        // as a scalar.
+        //
+        // I don't want to put this inside crh_ivk, but does it belong
+        // inside Scalar/Fr::from_bytes()? That seems the better
+        // place...
+        //
+        // https://github.com/zcash/librustzcash/blob/master/zcash_primitives/src/primitives.rs#L86
+        bytes[31] &= 0b0000_0111;
+
+        Self {
+            network: Network::default(),
+            scalar: Scalar::from_bytes(&bytes).unwrap(),
+        }
+    }
+}
+
+impl From<(AuthorizingKey, NullifierDerivingKey)> for IncomingViewingKey {
```

Comment:
@hdevalence Is this considered good style/practice? Seems ok to me but just wondering if there's a better way I'm not aware of.

Reply:
Hmm, I'm not sure.  `From` is sort of magical, which is nice when magic is desired.  Do we want magic here?

Also, another thing about `From` is that it works on owned values.  So unless AuthorizingKey, NullifierDerivingKey are Copy, using this conversion might require cloning the keys unnecessarily, unless the From impl was `impl<'a, 'b> From<(&'a AuthorizingKey, &'b NullifierDerivingKey)> ... ` which is a bit un-fun.

Reply:
That said, *I* like From impls, in general..... I think this is sort of an aesthetic choice